plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiSplitSource.java

```diff
@@ -47,30 +50,32 @@
         implements ConnectorSplitSource
 {
     private static final Logger log = Logger.get(HudiSplitSource.class);
-
+    private final Configuration conf;
     private final HudiTableHandle tableHandle;
     private final HoodieTableMetaClient metaClient;
     private final HoodieTableFileSystemView fileSystemView;
-    private final Iterator<HoodieBaseFile> hoodieBaseFileIterator;
+    private final Iterator<InputSplit> inputSplitIterator;
+    private final DataSize maxSplitSize;
     private final List<HivePartitionKey> partitionKeys;
     private final boolean metadataEnabled;
-    private final Optional<FileStatus[]> fileStatuses;
     private final String tablePath;
     private final String dataDir;
 
     public HudiSplitSource(
+            ConnectorSession session,
             HudiTableHandle tableHandle,
             Configuration conf,
             List<HivePartitionKey> partitionKeys,
             boolean metadataEnabled,
-            Optional<FileStatus[]> fileStatuses,
+            InputSplit[] inputSplits,
```

Comment:
Should we keep both fileStatuses and inputSplits? If we let Trino do the listing, then I think it would help the coordinator plan better. We can make it configurable and use splits for bootstrap and realtime.

Reply:
I can do that.

One thing I'd like to clarify.  So based on the implementation, if we use `HoodieParquetInputFormat`, the `HoodieParquetInputFormat::listStatus()` method gives us the file slices to be read, which incurs the loading of the active timeline (scanning file system to get instants and details) itself.  Similarly, `HoodieParquetInputFormat::getSplits()` split the files from `listStatus()`, given `getSplits()` calls `listStatus()` first internally.

In the current Hudi Trino connector draft, initializing file system view in the `HudiSplitSource` also loads the active timeline, incurring the overhead again.   There are a few ways we can avoid doubling the overhead:

(1) Completely rely on `HoodieParquetInputFormat::getSplits()` to get all splits, without using file system view in the `HudiSplitSource`.  It scans all instants for file slices and split all files synchronously before returning the results, blocking the processing.

(2) Rely on `HoodieParquetInputFormat::listStatus()` to get all file slices, and implement our own file split method (split logic in `HoodieParquetInputFormat::getSplits()` cannot be called indepedently from the class).  It still scans all instants for all file slices, but we can incrementally split each file upon batch request in `HudiSplitSource`.

(3) Do not use `HoodieParquetInputFormat`.  We let the Hudi Trino connector instantiates the file system view by loading the timeline, scanning the instants and details.  Then in the split source, we implement our own file split method.  Compare to (2), the mechanism is similar, with more code on the connector side.

I think we need to choose between (2) and (3).

Reply:
For now, I'll try (3) to refactor some of the logic so we have more control over how the file slices are scanned.

Reply:
(3) sounds good to me. One thing to note is that we initialize the metaclient just once per query. Within the session of a query, it's ok to list status for each partition but we should not be recreating meta client again and again.